### ISSUE TITLE
fast_matmul in docs example not working properly - returning wrong results

### ISSUE THREAD

**Issue by ALee008** [numba/numba#7324]
## fastmat_mul function from the docs returning wrong results

Hello,
I am using the latest version of numba, 0.53.1 as of now, and am working through the documentation *Numba for CUDA GPUs*. I copied the `fast_matmul` example documented [here](https://numba.pydata.org/numba-doc/dev/cuda/examples.html). But the result of a simple matrix multiplication is wrong. I also tried `matmul` function which is described on the same page and seems fine - for this simple example at least. The code I used is as follows, where `fast_matmul` and `matmul` are copied 1:1:

```python
import numpy as np
from numba import cuda, float32

# Controls threads per block and shared memory usage.
# The computation will be done on blocks of TPBxTPB elements.
TPB = 16

@cuda.jit
def fast_matmul(A, B, C):
    # Define an array in the shared memory
    # The size and type of the arrays must be known at compile time
    sA = cuda.shared.array(shape=(TPB, TPB), dtype=float32)
    sB = cuda.shared.array(shape=(TPB, TPB), dtype=float32)

    x, y = cuda.grid(2)

    tx = cuda.threadIdx.x
    ty = cuda.threadIdx.y
    bpg = cuda.gridDim.x    # blocks per grid

    if x >= C.shape[0] and y >= C.shape[1]:
        # Quit if (x, y) is outside of valid C boundary
        return

    # Each thread computes one element in the result matrix.
    # The dot product is chunked into dot products of TPB-long vectors.
    tmp = 0.
    for i in range(bpg):
        # Preload data into shared memory
        sA[tx, ty] = A[x, ty + i * TPB]
        sB[tx, ty] = B[tx + i * TPB, y]

        # Wait until all threads finish preloading
        cuda.syncthreads()

        # Computes partial product on the shared memory
        for j in range(TPB):
            tmp += sA[tx, j] * sB[j, ty]

        # Wait until all threads finish computing
        cuda.syncthreads()

    C[x, y] = tmp


@cuda.jit
def matmul(A, B, C):
    """Perform square matrix multiplication of C = A * B
    """
    i, j = cuda.grid(2)
    if i < C.shape[0] and j < C.shape[1]:
        tmp = 0.
        for k in range(A.shape[1]):
            tmp += A[i, k] * B[k, j]
        C[i, j] = tmp


if __name__ == '__main__':
    N, M = 2, 2
    A = np.arange(N * M).reshape(N, M)
    B = np.arange(N * M).reshape(N, M)
    C = np.empty_like(B)
    blocks_per_grid = (C.size + (TPB - 1)) // TPB
    fast_matmul[blocks_per_grid, TPB](A, B, C)
    print(C)
    print(np.dot(A, B))
    C = np.empty_like(A)
    blocks_per_grid = (C.size + (TPB - 1)) // TPB
    matmul[blocks_per_grid, TPB](A, B, C)
    print(C)
```
Expected result is 
```
[[ 2  3]
 [ 6 11]]
```
but I am getting 
```
[[    0 32762]
 [    0   502]]
```
Beware, these numbers can vary because matrix `C` is initialised using `np.empty_like()`.

Thank you and kind regards,
Ali

**Comment by gmarkall** [comment]
Thanks for the report - it does look like there's something wrong with the example. Although I don't reproduce the incorrect results with an RTX 8000, `compute-sanitizer` is not happy:

```
$ compute-sanitizer python repro.py 
========= COMPUTE-SANITIZER
/home/gmarkall/numbadev/numba/numba/cuda/compiler.py:872: NumbaPerformanceWarning: Grid size (1) < 2 * SM count (144) will likely result in GPU under utilization due to low occupancy.
  warn(NumbaPerformanceWarning(msg))
/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/devicearray.py:790: NumbaPerformanceWarning: Host array used in CUDA kernel will incur copy overhead to/from device.
  warn(NumbaPerformanceWarning(msg))
========= Invalid __global__ read of size 8 bytes
=========     at 0x460 in cudapy::__main__::fast_matmul$241(Array<long long, (int)2, C, mutable, aligned>, Array<long long, (int)2, C, mutable, aligned>, Array<long long, (int)2, C, mutable, aligned>)
=========     by thread (2,0,0) in block (0,0,0)
...
```

**Comment by gmarkall** [comment]
Resolved in #7330.
